src/core/preprocessor.js

```diff
@@ -44,12 +44,17 @@ class Preprocessor {
      * Run c-like preprocessor on the source code, and resolves the code based on the defines and ifdefs
      *
      * @param {string} source - The source code to work on.
-     * @param {Map<string, string>} [includes] - An object containing key-value pairs of include names and their
-     * content.
+     * @param {Map<string, string>} [defines] - A map containing key-value pairs of define names
+     * and their values. These are used for resolving #ifdef style of directives in the source.
+     * @param {Map<string, string>} [includes] - A map containing key-value pairs of include names
+     * and their content. These are used for resolving #include directives in the source.
      * @param {boolean} [stripUnusedColorAttachments] - If true, strips unused color attachments.
      * @returns {string|null} Returns preprocessed source code, or null in case of error.
      */
-    static run(source, includes = new Map(), stripUnusedColorAttachments = false) {
+    static run(source, defines, includes = new Map(), stripUnusedColorAttachments = false) {
```

Comment:
maybe also make `defines = new Map()` so you don't need the test below

Reply:
but then it nothing is passed in, I'd create a map and then clone it. I'd prefer to avoid it.